components/tikv_util/src/config.rs

```diff
@@ -457,17 +457,84 @@ impl<'de> Deserialize<'de> for ReadableDuration {
     }
 }
 
+fn canonicalize_imp<P: AsRef<Path>>(path: P) -> std::io::Result<PathBuf> {
+    fn normalize<P: AsRef<Path>>(path: P) -> PathBuf {
+        use std::path::Component;
+        let mut components = path.as_ref().components().peekable();
+        let mut ret = if let Some(c @ Component::Prefix(..)) = components.peek().cloned() {
+            components.next();
+            PathBuf::from(c.as_os_str())
+        } else {
+            PathBuf::new()
+        };
+
+        for component in components {
+            match component {
+                Component::Prefix(..) => unreachable!(),
+                Component::RootDir => {
+                    ret.push(component.as_os_str());
+                }
+                Component::CurDir => {}
+                Component::ParentDir => {
+                    ret.pop();
+                }
+                Component::Normal(c) => {
+                    ret.push(c);
+                }
+            }
+        }
+        ret
+    }
+    fn try_canonicalize_normalized_path<P: AsRef<Path>>(path: P) -> std::io::Result<PathBuf> {
+        use std::path::Component;
+        let mut components = path.as_ref().components().peekable();
+        let mut should_canonicalize = true;
+        let mut ret = if let Some(c @ Component::Prefix(..)) = components.peek().cloned() {
+            components.next();
+            PathBuf::from(c.as_os_str())
+        } else {
+            PathBuf::new()
+        };
```

Comment:
Should we canonicalize current directory first?

When we use a relative path and the first component of the path does not exist, this impl will not canonicalize as much as possible.

For example:

if current directory is '/folder1', and call `canonicalize_imp(PathBuf::from("./not-exists/not-exists-too")`, we will get output: `not-exists/not-exists-too`. It is supposed to give the output `/folder1/not-exists/not-exists-too`.